File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/network/JsonMessageSerializer.java
package com.github.dedis.popstellar.model.network.serializer.network;

import com.github.dedis.popstellar.model.network.method.*;
import com.github.dedis.popstellar.model.network.serializer.JsonUtils;
import com.google.gson.*;

import java.lang.reflect.Type;

import timber.log.Timber;

/** Json serializer and deserializer for the low level messages */
public class JsonMessageSerializer implements JsonSerializer<Message>, JsonDeserializer<Message> {

  private static final String TAG = JsonMessageSerializer.class.getSimpleName();

  @Override
  public Message deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
      throws JsonParseException {
    Timber.tag(TAG).d("deserializing message");
    JSONRPCRequest container = context.deserialize(json, JSONRPCRequest.class);
    JsonUtils.testRPCVersion(container.getJsonrpc());

    Method method = Method.find(container.getMethod());
    if (method == null) {
      throw new JsonParseException("Unknown method type " + container.getMethod());
    }
    JsonObject params = container.getParams();

    // If the Channeled Data is a Query, we need to give the params the id the the request
    if (method.expectResult()) {
      params.add(JsonUtils.JSON_REQUEST_ID, json.getAsJsonObject().get(JsonUtils.JSON_REQUEST_ID));
    }

    return context.deserialize(params, method.getDataClass());
  }

  @Override
  public JsonElement serialize(Message src, Type typeOfSrc, JsonSerializationContext context) {
    JsonObject params = context.serialize(src).getAsJsonObject();

    JsonObject obj =
        context
            .serialize(new JSONRPCRequest(JsonUtils.JSON_RPC_VERSION, src.getMethod(), params))
            .getAsJsonObject();

    if (src instanceof Query) {
      obj.addProperty(JsonUtils.JSON_REQUEST_ID, ((Query) src).getRequestId());
    }

    return obj;
  }
}

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/network/JsonMessageSerializer.kt
package com.github.dedis.popstellar.model.network.serializer.network

import com.github.dedis.popstellar.model.network.method.Message
import com.github.dedis.popstellar.model.network.method.Method
import com.github.dedis.popstellar.model.network.method.Query
import com.github.dedis.popstellar.model.network.serializer.JsonUtils
import com.github.dedis.popstellar.model.network.serializer.JsonUtils.testRPCVersion
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonParseException
import com.google.gson.JsonSerializationContext
import com.google.gson.JsonSerializer
import timber.log.Timber
import java.lang.reflect.Type

/** Json serializer and deserializer for the low level messages  */
class JsonMessageSerializer : JsonSerializer<Message>, JsonDeserializer<Message> {
    @Throws(JsonParseException::class)
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): Message {
        Timber.tag(TAG).d("deserializing message")
        val container = context.deserialize<JSONRPCRequest>(json, JSONRPCRequest::class.java)
        testRPCVersion(container.jsonrpc)
        val method = Method.find(container.method)
                ?: throw JsonParseException("Unknown method type " + container.method)
        val params = container.params

        // If the Channeled Data is a Query, we need to give the params the id the the request
        if (method.expectResult()) {
            params.add(JsonUtils.JSON_REQUEST_ID, json.asJsonObject[JsonUtils.JSON_REQUEST_ID])
        }
        return context.deserialize(params, method.dataClass)
    }

    override fun serialize(src: Message, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
        val params = context.serialize(src).asJsonObject
        val obj = context
                .serialize(JSONRPCRequest(JsonUtils.JSON_RPC_VERSION, src.method, params))
                .asJsonObject
        if (src is Query) {
            obj.addProperty(JsonUtils.JSON_REQUEST_ID, src.requestId)
        }
        return obj
    }

    companion object {
        private val TAG = JsonMessageSerializer::class.java.simpleName
    }
}